cmd/talosctl/pkg/talos/artifacts/copy.go

```diff
@@ -0,0 +1,222 @@
+			return fmt.Errorf("error mirroring image %d/%d (%q to %q): %w", i+1, len(images), srcImage, destImage, err)
+		}
+
+		srcSigRef, destSigRef, err := getSignature(ctx, options, ref, destRegistry)
+		if err != nil {
+			return fmt.Errorf("error getting signature for image %q (%d/%d): %w", srcImage, i+1, len(images), err)
+		}
+
+		srcSigRegistry := srcSigRef.Context().RegistryStr()
+		destSigRegistry := destSigRef.Context().RegistryStr()
+		sigRepo := srcSigRef.Context().RepositoryStr()
+		sigTag := extractTagAndDigest(srcSigRef.Name())
+
+		slog.Info("mirroring signature",
+			"src", srcSigRegistry,
+			"dest", destSigRegistry,
+			"image", sigRepo,
+			"tag", sigTag,
+			"progress", fmt.Sprintf("%d/%d", i+1, len(images)),
+		)
+
+		srcSig := srcSigRef.Name()
+		destSig := destSigRef.Name()
+
+		if err := copyManifest(ctx, options, srcSig, destSig); err != nil {
+			// We only log the error, not all images will have signatures
+			slog.Debug("error mirroring signature",
+				"error", err,
+				"src", srcSigRegistry,
+				"dest", destSigRegistry,
+				"image", sigRepo,
+				"tag", sigTag,
+				"progress", fmt.Sprintf("%d/%d", i+1, len(images)),
+			)
+		}
+	}
+
+	return nil
+}
+
+func extractTagAndDigest(image string) string {
+	// Find the repository/image separator
+	lastSlash := strings.LastIndex(image, "/")
+	if lastSlash == -1 {
+		lastSlash = -1
+	}
+
+	remainder := image[lastSlash+1:]
+
+	// Look for both tag and digest: name:tag@digest
+	if tagIdx := strings.Index(remainder, ":"); tagIdx != -1 {
+		if digestIdx := strings.Index(remainder, "@"); digestIdx != -1 {
+			// Both present: :tag@digest
+			return remainder[tagIdx:]
+		}
+		// Only tag: :tag
+		return remainder[tagIdx:]
+	}
+
+	// Only digest: @digest
+	if digestIdx := strings.Index(remainder, "@"); digestIdx != -1 {
+		return remainder[digestIdx:]
+	}
+
+	return ""
+}
+
+func rewriteRegistry(ref name.Reference, newRegistry string) (name.Reference, error) {
+	repo := ref.Context().RepositoryStr()
+
+	// Create new repository with the new registry
+	newRepo, err := name.NewRepository(newRegistry + "/" + repo)
+	if err != nil {
+		return nil, err
+	}
+
+	// Recreate the reference with the new repository
+	switch r := ref.(type) {
+	case name.Tag:
+		return name.NewTag(newRepo.String() + ":" + r.TagStr())
+	case name.Digest:
+		return name.NewDigest(newRepo.String() + "@" + r.DigestStr())
+	default:
+		return name.NewTag(newRepo.String() + ":latest")
+	}
+}
+
+func getSignature(ctx context.Context, options []crane.Option, ref name.Reference, destRegistry string) (name.Reference, name.Reference, error) {
+	r := retry.Exponential(
+		30*time.Second,
+		retry.WithUnits(time.Second),
+		retry.WithJitter(time.Second),
+		retry.WithErrorLogging(true),
+	)
+
+	var digest string
+
+	if err := r.RetryWithContext(
+		ctx, func(ctx context.Context) error {
+			options = append(options,
+				crane.WithContext(ctx),
+			)
+
+			var err error
+
+			digest, err = crane.Digest(
```

Comment:
I think we should find a signature for the exactly same digest as we mirror, to close the gap between resolving a tag and fetching a signature

Reply:
I think I've fixed it to what you've meant.